### ISSUE TITLE
Improve porting process

### ISSUE THREAD

**Issue by jodh-intel** [kata-containers/kata-containers#634]
As explained in [[1]], we need to improve our PR backporting (2.0 -> 1.x) and forward porting (1.x -> 2.0) process.

Even though we have the `port-to-2.0` label in all the 1.x repos, it's easy to either forget to add this label to a PR or forget to "follow through" and raise a new issue+PR in 2.0 for the forward port.

# Proposals

## Require maintainers to add new porting labels to PRs

Add the following pairs of labels to all the repositories:

- `needs-backport` and `no-backport-needed`
- `needs-forward-port` and `no-forward-port-needed`

Create a GitHub action to require **every** PR (*all* repos [1.x and 2.0]) to specify *one* of each pair of labels.

**Pros:**

+ Simple.
+ Should help to encourage more porting activity: the labels become a forcing function: maintainers are forced to consider whether a PR should be backported to 1.x (and/or the 1.x stable branch(es)) or forward ported to the 2.0 repo.

**Cons:**

- Extra process.
- Doesn't guarantee the porting activity the labels suggest will actually happen!

## Automatically raise porting issues

We could create GitHub actions to automatically create issues in the "other" repo (2.0 for 1.x PRs, and the appropriate 1.x repo for 2.0 PRs).

**Pros:**

+ Makes it harder to "forget" to do the porting work as there is a tracking issue to remind us

**Cons:**

- Still no guarantee the porting will take place.
- We will end up creating "too many" issues: some PRs won't need to be ported (however, better to raise one and then have it closed with a comment explaining *why* the port is not required!)

## Automatically raising speculative PRs

We could consider raising PRs on the "other" repo. We can then use the GitHub API to determine if these PRs are mergeable.

**Pros:**

+ This makes the porting as easy as possible: if the PR is mergeable, we have done most of the work for them to get the port landed.

**Cons:**

- We will end up raising "too many" PRs: some PRs won't need to be ported.

# See also

- https://github.com/kata-containers/community/wiki/Porting-PRs-between-Kata-1.x-and-Kata-2.0
- https://gist.github.com/jodh-intel/8686dd0ad04c9526cb5ba93283805561

---
[1]: http://lists.katacontainers.io/pipermail/kata-dev/2020-August/001468.html

**Comment by jodh-intel** [comment]
@kata-containers/architecture-committee - I've added this to the [agenda]( https://etherpad.opendev.org/p/Kata_Containers_2020_Architecture_Committee_Mtgs) for next week.

**Comment by jodh-intel** [comment]
# PR status

| Repo | PR |
|-|-|
| `agent` | https://github.com/kata-containers/agent/issues/827 |
| `ci` | https://github.com/kata-containers/ci/pull/323 |
| `community` | https://github.com/kata-containers/community/pull/169 |
| `documentation` | https://github.com/kata-containers/documentation/pull/726 |
| `kata-containers` | https://github.com/kata-containers/kata-containers/pull/640 |
| `ksm-throttler` | https://github.com/kata-containers/ksm-throttler/pull/166 |
| `osbuilder` | https://github.com/kata-containers/osbuilder/pull/492 |
| `packaging` | https://github.com/kata-containers/packaging/pull/1134 |
| `proxy` | https://github.com/kata-containers/proxy/pull/276 |
| `runtime` | https://github.com/kata-containers/runtime/pull/2913 |
| `shim` | https://github.com/kata-containers/shim/pull/269 |
| `tests` | https://github.com/kata-containers/tests/pull/2821 |
| `www.katacontainers.io` | https://github.com/kata-containers/www.katacontainers.io/pull/117 |

**Comment by jodh-intel** [comment]
Added dnm to all PRs except the proxy one, which we can use for testing again.

**Comment by jodh-intel** [comment]
Based on discussion on the gist in the description of this issue, here are the labels I think we need to require per PR:

| PR type | backport label | forward port label | Notes |
|-|-|-|-|
| PR does not need to be ported | `no-backport-needed` | `no-forward-port-needed` | A "standalone" PR. |
| PR that needs to be backported only | `needs-backport` | `no-forward-port-needed` |
| PR that needs to be forward ported only | `no-backport-needed` | `needs-forward-port` | |
| PR that needs to be ported backwards *and* forwards |  `needs-backport` |  `needs-forward-port` | For example a 1.x PR that needs to be ported to 2.0 *and* back to one or more 1.x stable branches. |
| A backport PR | `backport-pr` | | PR to actually make the backport changes. |
| A forward port PR | |  `forward-port-pr` | PR to actually make the forward port changes. |
| **Invalid scenario** | `backport-pr` | `forward-port-pr` | The GitHub action should disallow this combination of labels. |


In summary, if the PR is actually making a backport or forward-port, it only needs the appropriate label. But all other PRs will need two labels.

/cc @ariel-adam, @c3d, @amshinde, @gnawux.

**Comment by jodh-intel** [comment]
I've now added the `backport-pr` and `forward-port-pr` labels to all Kata repos.

**Comment by amshinde** [comment]
@jodh-intel Thanks for adding the labels. Makes sense to me.  I have acked the PRs that you have raised for this. 
@c3d @ariel-adam @gnawux PTAL

**Comment by jodh-intel** [comment]
More tweaks: I've changed the [table above](https://github.com/kata-containers/kata-containers/issues/634#issuecomment-688439361) to specify `backport` and `forward-port` labels (rather than `backport-pr` and `forward-port-pr`), and updated all associated PRs and labels in all repos.

The rationale for this being that by making the label names more generic (removing "`pr`"), they can also be used for issues too.

This would allow us to make the porting process even more robust when we're at the point we can auto-raise issues in the "other" repo since we could use a combination of `hub` and [`hub-util.sh`](https://github.com/kata-containers/.github/tree/master/scripts) to:

- List issues in the "other" repo which are labeled `backport` or `forward-port`.
- Check to ensure each porting issue has a corresponding porting PR.
- Potentially block a release until the above is true.

**Comment by jodh-intel** [comment]
Alas, we're going to need another round of PRs to change each repos workflow file to call the script on https://github.com/kata-containers/.github/pull/21, rather than embedding the shell script in the YAML as is currently done. Why? Read: https://github.com/kata-containers/.github/issues/20.

**Comment by jodh-intel** [comment]
Updated process docs on https://github.com/kata-containers/community/pull/178.

**Comment by fidencio** [comment]
I think we've reached a reasonable state of the "porting process".
@jodh-intel, I'm closing the issue but feel free to re-open and update the issue accordingly in case you feel the need to do so.
